From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: celenity <celenity@celenity.dev>
Date: Sun, 28 Jun 2026 05:34:22 +0000
Subject: [PATCH] Fenix - Local wallpapers

Adds support for local wallpapers, to prevent downloading them from Mozilla servers upon each launch.

Signed-off-by: celenity <celenity@celenity.dev>
---
diff --git a/mobile/android/fenix/app/src/main/java/org/mozilla/fenix/components/UseCases.kt b/mobile/android/fenix/app/src/main/java/org/mozilla/fenix/components/UseCases.kt
index 24554c7ab197..e238a030d3e1 100644
--- a/mobile/android/fenix/app/src/main/java/org/mozilla/fenix/components/UseCases.kt
+++ b/mobile/android/fenix/app/src/main/java/org/mozilla/fenix/components/UseCases.kt
@@ -150,6 +150,7 @@ class UseCases(
             rootStorageDirectory to currentLocale
         }
         WallpapersUseCases(
+            context.resources,
             context.components.settings,
             rootStorageDirectory,
             appStore.value,
diff --git a/mobile/android/fenix/app/src/main/java/org/mozilla/fenix/settings/wallpaper/WallpaperSettings.kt b/mobile/android/fenix/app/src/main/java/org/mozilla/fenix/settings/wallpaper/WallpaperSettings.kt
index ace40a777e07..e356264082aa 100644
--- a/mobile/android/fenix/app/src/main/java/org/mozilla/fenix/settings/wallpaper/WallpaperSettings.kt
+++ b/mobile/android/fenix/app/src/main/java/org/mozilla/fenix/settings/wallpaper/WallpaperSettings.kt
@@ -58,6 +58,7 @@ import org.mozilla.fenix.theme.FirefoxTheme
 import org.mozilla.fenix.theme.PreviewThemeProvider
 import org.mozilla.fenix.theme.Theme
 import org.mozilla.fenix.wallpapers.Wallpaper
+import org.ironfoxoss.core.R as ironfoxR
 
 /**
  * The screen for controlling settings around Wallpapers. When a new wallpaper is selected,
@@ -145,7 +146,7 @@ private fun WallpaperGroupHeading(
             modifier = Modifier.semantics(mergeDescendants = true, properties = headingSemantics),
         ) {
             Text(
-                text = stringResource(R.string.wallpaper_artist_series_title),
+                text = stringResource(ironfoxR.string.wallpaper_collection_fennec),
                 color = MaterialTheme.colorScheme.onSurface,
                 style = FirefoxTheme.typography.subtitle2,
             )
@@ -159,7 +160,7 @@ private fun WallpaperGroupHeading(
                 )
             } else {
                 val link = stringResource(R.string.wallpaper_learn_more)
-                val text = stringResource(R.string.wallpaper_artist_series_description_with_learn_more, link)
+                val text = stringResource(ironfoxR.string.wallpaper_collection_fennec_description, link)
                 val linkStartIndex = text.indexOf(link)
                 val linkEndIndex = linkStartIndex + link.length
 
diff --git a/mobile/android/fenix/app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperMetadataFetcher.kt b/mobile/android/fenix/app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperMetadataFetcher.kt
index ab7ba29ba34c..1abb54b8c61e 100644
--- a/mobile/android/fenix/app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperMetadataFetcher.kt
+++ b/mobile/android/fenix/app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperMetadataFetcher.kt
@@ -6,11 +6,11 @@ package org.mozilla.fenix.wallpapers
 
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.withContext
-import mozilla.components.concept.fetch.Client
-import mozilla.components.concept.fetch.Request
+import android.content.res.Resources
+import org.ironfoxoss.core.R
 import org.json.JSONArray
 import org.json.JSONObject
-import org.mozilla.fenix.BuildConfig
+// import org.mozilla.fenix.BuildConfig
 import org.mozilla.fenix.utils.toHexColor
 import java.text.SimpleDateFormat
 import java.util.Date
@@ -22,23 +22,18 @@ import java.util.Locale
  * @param client The client that will be used to fetch metadata.
  */
 class WallpaperMetadataFetcher(
-    private val client: Client,
+    private val resources: Resources,
 ) {
-    private val metadataUrl = BuildConfig.WALLPAPER_URL.substringBefore("android") +
-        "metadata/v$CURRENT_JSON_VERSION/wallpapers.json"
+//    private val metadataUrl = BuildConfig.WALLPAPER_URL.substringBefore("android") +
+//        "metadata/v$CURRENT_JSON_VERSION/wallpapers.json"
 
     /**
      * Downloads the list of wallpapers from the remote source. Failures will return an empty list.
      */
     suspend fun downloadWallpaperList(): List<Wallpaper> = withContext(Dispatchers.IO) {
-        Result.runCatching {
-            val request = Request(url = metadataUrl, method = Request.Method.GET, conservative = true)
-            val response = client.fetch(request)
-            response.body.useBufferedReader {
-                val json = it.readText()
-                JSONObject(json).parseAsWallpapers()
-            }
-        }.getOrElse { listOf() }
+        val json = resources.openRawResource(R.raw.wallpapers).bufferedReader()
+            .use { it.readText() }
+        return@withContext JSONObject(json).parseAsWallpapers()
     }
 
     private fun JSONObject.parseAsWallpapers(): List<Wallpaper> =
diff --git a/mobile/android/fenix/app/src/main/java/org/mozilla/fenix/wallpapers/WallpapersUseCases.kt b/mobile/android/fenix/app/src/main/java/org/mozilla/fenix/wallpapers/WallpapersUseCases.kt
index d11145f10a38..185b1a955ca3 100644
--- a/mobile/android/fenix/app/src/main/java/org/mozilla/fenix/wallpapers/WallpapersUseCases.kt
+++ b/mobile/android/fenix/app/src/main/java/org/mozilla/fenix/wallpapers/WallpapersUseCases.kt
@@ -17,6 +17,8 @@ import org.mozilla.fenix.components.appstate.AppAction
 import org.mozilla.fenix.utils.Settings
 import java.io.File
 import java.util.Date
+import android.content.res.Resources
+import org.ironfoxoss.ironfox.utils.IronFoxWallpaperUtils
 
 /**
  * Contains use cases related to the wallpaper feature.
@@ -29,6 +31,7 @@ import java.util.Date
  * @param currentLocale The locale currently being used on the device.
  */
 class WallpapersUseCases(
+    resources: Resources,
     settings: Settings,
     filesDir: File,
     appStore: AppStore,
@@ -46,7 +49,7 @@ class WallpapersUseCases(
     // Use case for initializing wallpaper feature. Should usually be called early
     // in the app's lifetime to ensure that any potential long-running tasks can complete quickly.
     val initialize: InitializeWallpapersUseCase by lazy {
-        val metadataFetcher = WallpaperMetadataFetcher(client)
+        val metadataFetcher = WallpaperMetadataFetcher(resources)
         val migrationHelper = LegacyWallpaperMigration(
             storageRootDirectory = storageRootDirectory,
             settings = settings,
@@ -66,12 +69,12 @@ class WallpapersUseCases(
     // Use case for loading specific wallpaper bitmaps.
     val loadBitmap: LoadBitmapUseCase by lazy {
         DefaultLoadBitmapUseCase(
-            getFilesDir = { filesDir },
+            resources = resources
         )
     }
 
     val loadThumbnail: LoadThumbnailUseCase by lazy {
-        DefaultLoadThumbnailUseCase(storageRootDirectory)
+        DefaultLoadThumbnailUseCase(resources = resources)
     }
 
     // Use case for selecting a new wallpaper.
@@ -196,7 +199,7 @@ class WallpapersUseCases(
 
     @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
     internal class DefaultLoadBitmapUseCase(
-        private val getFilesDir: suspend () -> File,
+        private val resources: Resources
     ) : LoadBitmapUseCase {
         override suspend fun invoke(wallpaper: Wallpaper, orientation: Int): Bitmap? =
             loadWallpaperFromDisk(wallpaper, orientation)
@@ -206,10 +209,10 @@ class WallpapersUseCases(
             wallpaper: Wallpaper,
             orientation: Int,
         ): Bitmap? = try {
-            val path = wallpaper.getLocalPathFromContext(orientation)
+            val drawableRes = if (wallpaper.name !== Wallpaper.DEFAULT) IronFoxWallpaperUtils.getWallpaperDrawable(wallpaper.name, orientation) else return null
             withContext(Dispatchers.IO) {
-                val file = File(getFilesDir(), path)
-                BitmapFactory.decodeStream(file.inputStream())
+//                val file = File(getFilesDir(), path)
+                BitmapFactory.decodeResource(resources, drawableRes)
             }
         } catch (e: CancellationException) {
             // CancellationException must not be swallowed: if the coroutine was canceled while loading,
@@ -247,13 +250,13 @@ class WallpapersUseCases(
     }
 
     @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
-    internal class DefaultLoadThumbnailUseCase(private val filesDir: File) : LoadThumbnailUseCase {
+    internal class DefaultLoadThumbnailUseCase(private val resources: Resources) : LoadThumbnailUseCase {
         override suspend fun invoke(wallpaper: Wallpaper): Bitmap? = withContext(Dispatchers.IO) {
             Result.runCatching {
-                val path = Wallpaper.getLocalPath(wallpaper.name, Wallpaper.ImageType.Thumbnail)
+                val drawableRes = if (wallpaper.name !== Wallpaper.DEFAULT) IronFoxWallpaperUtils.getWallpaperDrawableThumbnail(wallpaper.name) else return@runCatching null
                 withContext(Dispatchers.IO) {
-                    val file = File(filesDir, path)
-                    BitmapFactory.decodeStream(file.inputStream())
+//                    val file = File(filesDir, path)
+                    BitmapFactory.decodeResource(resources, drawableRes)
                 }
             }.getOrNull()
         }
@@ -285,7 +288,7 @@ class WallpapersUseCases(
          */
         override suspend fun invoke(wallpaper: Wallpaper): Wallpaper.ImageFileState {
             return if (wallpaper.collection == Wallpaper.DefaultCollection ||
-                fileManager.wallpaperImagesExist(wallpaper)
+                true
             ) {
                 selectWallpaper(wallpaper)
                 dispatchDownloadState(wallpaper, Wallpaper.ImageFileState.Downloaded)
-- 
